src/LogWindow.cpp

```diff
@@ -0,0 +1,229 @@
+         sFrame.get());
+      return;
+   }
+}
+
+void LogWindowListener::UpdatePrefs()
+{
+   //! Re-create the non-modal window in case of change of preferred language
+   if (sFrame) {
+      bool shown = sFrame->IsShown();
+      if (shown) {
+         LogWindow::Show(false);
+      }
+      sFrame.reset();
+      if (shown) {
+         LogWindow::Show(true);
+      }
+   }
+}
+}
+
+void LogWindow::Show(bool show)
+{
+   // Hide the frame if created, otherwise do nothing
+   if (!show) {
+      if (sFrame) {
+         sFrame->Show(false);
+      }
+      return;
+   }
+
+   // If the frame already exists, refresh its contents and show it
+   auto pLogger = AudacityLogger::Get();
+   if (sFrame) {
+      if (!sFrame->IsShown() && sText) {
+         if (pLogger)
+            sText->ChangeValue(pLogger->GetBuffer());
+         sText->SetInsertionPointEnd();
+         sText->ShowPosition(sText->GetLastPosition());
+      }
+      sFrame->Show();
+      sFrame->Raise();
+      return;
+   }
+
+   // This is the first use, so create the frame
+   Destroy_ptr<wxFrame> frame
+      { safenew wxFrame(NULL, wxID_ANY, _("Audacity Log")) };
+   frame->SetName(frame->GetTitle());
+   frame->SetBackgroundColour(wxSystemSettings::GetColour(wxSYS_COLOUR_3DFACE));
+
+   // loads either the XPM or the windows resource, depending on the platform
+   {
+#if !defined(__WXMAC__) && !defined(__WXX11__)
+#if defined(__WXMSW__)
+      wxIcon ic{wxICON(AudacityLogo)};
+#elif defined(__WXGTK__)
+      wxIcon ic{wxICON(AudacityLogoAlpha)};
+#else
+      wxIcon ic{};
+      ic.CopyFromBitmap(theTheme.Bitmap(bmpAudacityLogo48x48));
+#endif
+      frame->SetIcon(ic);
+#endif
+   }
+
+   // Log text
+   ShuttleGui S(frame.get(), eIsCreating);
+
+   S.Style(wxNO_BORDER | wxTAB_TRAVERSAL).Prop(true).StartPanel();
+   {
+      S.StartVerticalLay(true);
+      {
+         sText = S.Style(wxTE_MULTILINE | wxHSCROLL | wxTE_READONLY | wxTE_RICH)
+            .AddTextWindow(pLogger ? pLogger->GetBuffer() : wxString{});
+
+         S.AddSpace(0, 5);
+         S.StartHorizontalLay(wxALIGN_CENTER, 0);
+         {
+            S.AddSpace(10, 0);
+            S.Id(LoggerID_Save).AddButton(XXO("&Save..."));
+            S.Id(LoggerID_Clear).AddButton(XXO("Cl&ear"));
+            S.Id(LoggerID_Close).AddButton(XXO("&Close"));
+            S.AddSpace(10, 0);
+         }
+         S.EndHorizontalLay();
+         S.AddSpace(0, 3);
+      }
+      S.EndVerticalLay();
+   }
+   S.EndPanel();
+
+   // Give a place for the menu help text to go
+   // frame->CreateStatusBar();
+
+   frame->Layout();
+
+   // Hook into the frame events
+   frame->Bind(wxEVT_CLOSE_WINDOW, OnCloseWindow );
+
+   frame->Bind(   wxEVT_COMMAND_MENU_SELECTED, OnSave, LoggerID_Save);
+   frame->Bind(   wxEVT_COMMAND_MENU_SELECTED, OnClear, LoggerID_Clear);
+   frame->Bind(   wxEVT_COMMAND_MENU_SELECTED, OnClose, LoggerID_Close);
+   frame->Bind(   wxEVT_COMMAND_BUTTON_CLICKED, OnSave, LoggerID_Save);
```

Comment:
This moment yet not clearly, to bind several methods (`OnSave`, ...) to one single event `wxEVT_COMMAND_BUTTON_CLICKED`, it will not overrided?